File: mobile/ios/Air/SubModules/AirAsFramework/Deeplink/DeeplinkParser.swift

import Foundation
import UIInAppBrowser
import WalletCore
import WalletContext

extension Deeplink {
    init?(url: URL) {
        let deeplink: Deeplink? = switch url.scheme {
        case "ton":
            parseTonInvoiceUrl(url)
        case "tc", "mytonwallet-tc":
            parseTonConnectUrl(url)
        case "wc":
            parseWalletConnectUrl(url)
        case "mtw":
            parseMtwUrl(url)
        case "https", "http":
            switch url.host {
            case "connect.mytonwallet.org":
                parseTonConnectUrl(url)
            case "walletconnect.com":
                if url.path == "/wc" {
                    parseWalletConnectUrl(url)
                } else {
                    nil
                }
            case "go.mytonwallet.org":
                parseMtwUrl(url)
            case "my.tt":
                parseMtwUrl(url)
            default:
                nil
            }
        default:
            nil
        }
        if let deeplink {
            self = deeplink
        } else {
            return nil
        }
    }
}

private func parseTonConnectUrl(_ url: URL) -> Deeplink {
    Deeplink.tonConnect2(requestLink: url.absoluteString)
}

private func parseWalletConnectUrl(_ url: URL) -> Deeplink {
    Deeplink.walletConnect(requestLink: url.absoluteString)
}

private func parseTonInvoiceUrl(_ url: URL) -> Deeplink? {
    guard let parsedWalletURL = parseTonTransferUrl(url) else {
        return nil
    }
    return Deeplink.invoice(
        address: parsedWalletURL.address,
        amount: parsedWalletURL.amount,
        comment: parsedWalletURL.comment,
        binaryPayload: parsedWalletURL.bin,
        token: parsedWalletURL.token,
        jetton: parsedWalletURL.jetton,
        stateInit: parsedWalletURL.stateInit
    )
}

private func parseMtwUrl(_ url: URL) -> Deeplink? {
    
    var url = url
    
    if url.scheme == "http", var components = URLComponents(url: url, resolvingAgainstBaseURL: false) {
        components.scheme = "https"
        if let newUrl = components.url {
            url = newUrl
        }
    }
    
    if url.scheme == "https" {
        let urlString = url.absoluteString
        for universalUrl in SELF_UNIVERSAL_URLS {
            if urlString.starts(with: universalUrl) {
                let newUrlString = urlString.replacing(universalUrl, with: SELF_PROTOCOL, maxReplacements: 1)
                if let newUrl = URL(string: newUrlString) {
                    url = newUrl
                }
                break
            }
        }
    }
    
    switch url.host {
    case "classic":
        return .switchToClassic
        
    case "swap", "buy-with-crypto":
        var from: String? = nil
        var to: String? = nil
        var amountIn: Double? = nil
        if let query = url.query, let components = URLComponents(string: "/?" + query), let queryItems = components.queryItems {
            for queryItem in queryItems {
                if let value = queryItem.value {
                    if queryItem.name == "amountIn", !value.isEmpty, let amountValue = Double(value) {
                        amountIn = amountValue
                    } else if queryItem.name == "in", !value.isEmpty {
                        from = value
                    } else if queryItem.name == "out", !value.isEmpty {
                        to = value
                    }
                }
            }
        }
        if url.host == "buy-with-crypto" {
            if to == nil, from != "toncoin" {
                to = "toncoin"
            }
            if from == nil {
                from = TRON_USDT_SLUG
            }
        }
        return .swap(from: from, to: to, amountIn: amountIn)

    case "transfer":
        if url.pathComponents.count <= 1 {
            return .transfer
        } else {
            return parseTonInvoiceUrl(url)
        }

    case "buy-with-card":
        return .buyWithCard
        
    case Deeplink.Sell.urlHost:
        return .sell(.init(url))

    case "stake":
        return .stake

    case "giveaway":
        let pathname = url.absoluteString
        let regex = try! NSRegularExpression(pattern: "giveaway/([^/]+)")
        var giveawayId: String? = nil
        
        if let match = regex.firstMatch(in: pathname, range: NSRange(pathname.startIndex..., in: pathname)) {
            let giveawayIdRange = match.range(at: 1)
            if let giveawayIdRange = Range(giveawayIdRange, in: pathname) {
                giveawayId = String(pathname[giveawayIdRange])
            }
        }
        let urlString = "https://giveaway.mytonwallet.io/\(giveawayId != nil ? "?giveawayId=\(giveawayId!)" : "")"
        let url = URL(string: urlString)!
        return .url(config: InAppBrowserPageConfig(
                url: url,
                title: "Giveaway",
                injectDappConnect: true
        ))
        
    case "r":
        let pathname = url.absoluteString
        let regex = try! NSRegularExpression(pattern: "r/([^/]+)")
        var r: String? = nil
        
        if let match = regex.firstMatch(in: pathname, range: NSRange(pathname.startIndex..., in: pathname)) {
            let rIdRange = match.range(at: 1)
            if let rRange = Range(rIdRange, in: pathname) {
                r = String(pathname[rRange])
            }
        }
        let urlString = "https://checkin.mytonwallet.org/\(r != nil ? "?r=\(r!)" : "")"
        let url = URL(string: urlString)!
        return .url(config: InAppBrowserPageConfig(
            url: url,
            title: "Checkin",
            injectDappConnect: true
        ))
        
    case "receive":
        return .receive
        
    case "explore":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        return .explore(siteHost: pathComponents.first?.nilIfEmpty)

    case "token":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        if pathComponents.count >= 2 {
            // mtw://token/{chain}/{tokenAddress}
            guard let chain = ApiChain(rawValue: pathComponents[0]) else { return nil }
            let tokenAddress = pathComponents[1]
            return .tokenAddress(chain: chain, tokenAddress: tokenAddress)
        } else if pathComponents.count == 1 {
            // mtw://token/{slug}
            let slug = pathComponents[0]
            return .tokenSlug(slug: slug)
        } else {
            return nil
        }
    
    case "tx":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard pathComponents.count >= 2 else { return nil }
        let chainString = pathComponents[0]
        let txId = pathComponents.dropFirst().joined(separator: "/")
        guard let chain = ApiChain(rawValue: chainString), chain.isSupported else { return nil }
        return .transaction(chain: chain, txId: txId)

    case "nft":
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard let nftAddress = pathComponents.first?.nilIfEmpty else { return nil }
        return .nftAddress(nftAddress: nftAddress)
        
    case "view":
        var addressOrDomainByChain: [String: String] = [:]
        let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        for queryItem in queryItems {
            if let chain = ApiChain(rawValue: queryItem.name),
               chain.isSupported,
               let addressOrDomain = queryItem.value?.nilIfEmpty,
               chain.isValidAddressOrDomain(addressOrDomain) {
                addressOrDomainByChain[chain.rawValue] = addressOrDomain
            }
        }
        return .view(addressOrDomainByChain: addressOrDomainByChain)

    default:
        return nil
    }
}

extension Deeplink {
    struct Sell {
        static let urlHost = "offramp"
        
        let transactionId: String?
        let baseCurrencyCode: String?
        let baseCurrencyAmount: String?
        let depositWalletAddress: String?
        let depositWalletAddressTag: String?

        init(_ url: URL) {
            let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
            var transactionId: String?
            var baseCurrencyCode: String?
            var baseCurrencyAmount: String?
            var depositWalletAddress: String?
            var depositWalletAddressTag: String?

            for item in queryItems {
                guard let value = item.value, !value.isEmpty else { continue }
                switch item.name {
                case "transactionId": transactionId = value
                case "baseCurrencyCode": baseCurrencyCode = value
                case "baseCurrencyAmount": baseCurrencyAmount = value
                case "depositWalletAddress": depositWalletAddress = value
                case "depositWalletAddressTag": depositWalletAddressTag = value
                default: break
                }
            }
            
            self.transactionId = transactionId
            self.baseCurrencyCode = baseCurrencyCode
            self.baseCurrencyAmount = baseCurrencyAmount
            self.depositWalletAddress = depositWalletAddress
            self.depositWalletAddressTag = depositWalletAddressTag
        }
    }
}

File: mobile/android/air/SubModules/WalletCore/src/main/java/org/mytonwallet/app_air/walletcore/deeplink/DeeplinkParser.kt
package org.mytonwallet.app_air.walletcore.deeplink

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import org.mytonwallet.app_air.walletcontext.helpers.AddressHelpers
import org.mytonwallet.app_air.walletcontext.models.MBlockchainNetwork
import org.mytonwallet.app_air.walletcore.TRON_USDT_SLUG
import org.mytonwallet.app_air.walletcore.models.InAppBrowserConfig
import org.mytonwallet.app_air.walletcore.models.blockchain.MBlockchain
import java.net.URLDecoder

sealed class Deeplink {
    abstract val accountAddress: String?

    data class TonConnect2(override val accountAddress: String?, val requestUri: Uri) : Deeplink() {
        val isConnectRequest: Boolean
            get() {
                return !requestUri.getQueryParameter("r").isNullOrBlank()
            }
    }

    data class Invoice(
        override val accountAddress: String?,
        val address: String,
        val amount: String?,
        val binary: String?,
        val expiry: Long?,
        val init: String?,
        val jetton: String?,
        val comment: String?,
        val token: String?,
        val hasUnsupportedParams: Boolean
    ) : Deeplink()

    data class Swap(
        override val accountAddress: String?,
        val from: String?,
        val to: String?,
        val amountIn: Double?
    ) : Deeplink()

    data class Receive(override val accountAddress: String?) : Deeplink()
    data class BuyWithCard(override val accountAddress: String?) : Deeplink()
    data class Offramp(
        override val accountAddress: String?,
        val transactionId: String?,
        val baseCurrencyCode: String?,
        val baseCurrencyAmount: String?,
        val depositWalletAddress: String?,
        val depositWalletAddressTag: String?
    ) : Deeplink()

    data class Stake(override val accountAddress: String?) : Deeplink()
    data class Explore(
        override val accountAddress: String?,
        val targetUri: Uri?
    ) : Deeplink()

    data class Url(
        override val accountAddress: String?,
        val config: InAppBrowserConfig
    ) : Deeplink()

    data class Transaction(
        override val accountAddress: String?,
        val chain: String?,
        val txId: String?,
        val txHash: String?,
        val isPushNotification: Boolean
    ) : Deeplink()

    data class TokenBySlug(override val accountAddress: String?, val slug: String) : Deeplink()
    data class TokenByAddress(
        override val accountAddress: String?,
        val chain: String,
        val address: String
    ) : Deeplink()

    data class StakeTx(override val accountAddress: String?) : Deeplink()
    data class ExpiringDns(override val accountAddress: String?, val domainAddress: String) :
        Deeplink()

    data class WalletConnect(override val accountAddress: String?, val requestUri: Uri) : Deeplink()
    data class SwitchToLegacy(override val accountAddress: String?) : Deeplink()
    data class View(
        override val accountAddress: String?,
        val network: MBlockchainNetwork,
        val addressByChain: Map<String, String>
    ) : Deeplink()

    data class Nft(
        override val accountAddress: String?,
        val network: MBlockchainNetwork,
        val nftAddress: String
    ) : Deeplink()
}

interface DeeplinkNavigator {
    fun handle(deeplink: Deeplink)
}

class DeeplinkParser {

    companion object {
        fun parse(intent: Intent): Deeplink? {
            /*Logger.d(Logger.LogTag.DEEPLINK, "Data ${intent.data}")
            intent.extras?.keySet()?.forEach { key ->
                Logger.d(Logger.LogTag.DEEPLINK, "Extra $key: ${intent.extras?.getString(key, "")}")
            }*/
            return parse(intent.data) ?: parse(intent.extras)
        }

        fun parse(uri: Uri?): Deeplink? {
            if (uri == null)
                return null
            return when (uri.scheme) {
                "ton" -> handleTonInvoice(uri)
                "tc", "mytonwallet-tc" -> handleTonConnect(uri)
                "wc" -> handleWalletConnect(uri)
                "mtw" -> handleMTW(uri)
                "https" -> handleHttpsDeeplinks(uri)
                else -> {
                    null
                }
            }
        }

        private fun parse(bundle: Bundle?): Deeplink? {
            if (bundle == null)
                return null
            val address = bundle.getString("address") ?: return null
            return when (bundle.getString("action")) {
                "openUrl" -> {
                    val url = bundle.getString("url") ?: return null
                    Deeplink.Url(
                        address, InAppBrowserConfig(
                            url = url,
                            title = bundle.getString("title"),
                            injectDappConnect = true
                        )
                    )
                }

                "nativeTx", "swap", "jettonTx" -> {
                    val txId = bundle.getString("txId") ?: return null
                    val chain = bundle.getString("chain") ?: MBlockchain.ton.name
                    return Deeplink.Transaction(
                        accountAddress = address,
                        chain = chain,
                        txId = txId,
                        txHash = null,
                        isPushNotification = true
                    )
                }

                "staking" -> {
                    return Deeplink.StakeTx(accountAddress = address)
                }

                "expiringDns" -> {
                    val domainAddress = bundle.getString("domainAddress") ?: return null
                    return Deeplink.ExpiringDns(
                        accountAddress = address,
                        domainAddress = domainAddress
                    )
                }

                else -> {
                    return null
                }
            }
        }

        private fun handleTonConnect(uri: Uri): Deeplink {
            return Deeplink.TonConnect2(accountAddress = null, requestUri = uri)
        }

        private fun handleWalletConnect(uri: Uri): Deeplink {
            return Deeplink.WalletConnect(accountAddress = null, requestUri = uri)
        }

        private fun handleTonInvoice(uri: Uri): Deeplink? {
            val parsedWalletURL = parseWalletUrl(uri) ?: return null
            return Deeplink.Invoice(
                accountAddress = null,
                address = parsedWalletURL.address,
                amount = parsedWalletURL.amount,
                binary = parsedWalletURL.binary,
                comment = parsedWalletURL.comment,
                expiry = parsedWalletURL.expiry,
                init = parsedWalletURL.init,
                jetton = parsedWalletURL.jetton,
                token = parsedWalletURL.token,
                hasUnsupportedParams = parsedWalletURL.hasUnsupportedParams,
            )
        }

        private fun handleHttpsDeeplinks(uri: Uri): Deeplink? {
            when (uri.host) {
                "my.tt", "go.mytonwallet.org" -> {
                    val path = uri.path?.trimStart('/') ?: ""
                    val pathSegments = path.split('/')

                    val mtwUri = Uri.Builder()
                        .scheme("mtw")
                        .authority(pathSegments.firstOrNull() ?: "")
                        .apply {
                            if (pathSegments.size > 1) {
                                pathSegments.drop(1).forEach { segment ->
                                    appendPath(segment)
                                }
                            }
                        }
                        .encodedQuery(uri.encodedQuery)
                        .build()
                    return handleMTW(mtwUri)
                }

                "connect.mytonwallet.org" -> {
                    return handleTonConnect(uri)
                }

                "walletconnect.com" -> {
                    if (uri.path == "/wc") return handleWalletConnect(uri)
                }
            }
            return null
        }

        private fun handleMTW(uri: Uri): Deeplink? {
            return when (uri.host) {
                "swap", "buy-with-crypto" -> {
                    var from: String? = null
                    var to: String? = null
                    var amountIn: Double? = null

                    uri.query?.let { query ->
                        val components = URLDecoder.decode(query, "UTF-8").split("&").mapNotNull {
                            it.split("=")
                                .let { parts -> if (parts.size == 2) parts[0] to parts[1] else null }
                        }.toMap()

                        components["amountIn"]?.toDoubleOrNull()?.let { amountIn = it }
                        components["in"]?.let { from = it }
                        components["out"]?.let { to = it }
                    }

                    if (uri.host == "buy-with-crypto") {
                        if (to == null && from != "toncoin") to = "toncoin"
                        if (from == null) from = TRON_USDT_SLUG
                    }

                    Deeplink.Swap(accountAddress = null, from = from, to = to, amountIn = amountIn)
                }

                "transfer" -> handleTonInvoice(uri)
                "receive" -> Deeplink.Receive(accountAddress = null)
                "buy-with-card" -> Deeplink.BuyWithCard(accountAddress = null)
                "offramp" -> Deeplink.Offramp(
                    accountAddress = null,
                    transactionId = uri.getQueryParameter("transactionId"),
                    baseCurrencyCode = uri.getQueryParameter("baseCurrencyCode"),
                    baseCurrencyAmount = uri.getQueryParameter("baseCurrencyAmount"),
                    depositWalletAddress = uri.getQueryParameter("depositWalletAddress"),
                    depositWalletAddressTag = uri.getQueryParameter("depositWalletAddressTag")
                )

                "stake" -> Deeplink.Stake(accountAddress = null)
                "explore" -> Deeplink.Explore(
                    accountAddress = null,
                    targetUri = uri.extractSubUri()
                )

                "giveaway" -> {
                    val giveawayId = extractId(uri.toString(), "giveaway/([^/]+)")
                    val urlString =
                        "https://giveaway.mytonwallet.io/" + if (giveawayId != null) "?giveawayId=$giveawayId" else ""
                    val config = InAppBrowserConfig(
                        url = urlString,
                        title = "Giveaway",
                        injectDappConnect = true
                    )
                    Deeplink.Url(accountAddress = null, config)
                }

                "r" -> {
                    val rId = extractId(uri.toString(), "r/([^/]+)")
                    val urlString =
                        "https://checkin.mytonwallet.org/" + if (rId != null) "?r=$rId" else ""
                    val config = InAppBrowserConfig(
                        url = urlString,
                        title = "Checkin",
                        injectDappConnect = true
                    )
                    Deeplink.Url(accountAddress = null, config)
                }

                "classic" -> {
                    Deeplink.SwitchToLegacy(null)
                }

                "token" -> {
                    val pathParts = uri.pathSegments

                    if (pathParts.size > 1) {
                        val chain = pathParts[0]
                        val tokenAddress = pathParts[1]
                        return Deeplink.TokenByAddress(null, chain, tokenAddress)
                    } else {
                        pathParts.firstOrNull()?.let { tokenSlug ->
                            return Deeplink.TokenBySlug(null, tokenSlug)
                        }
                    }
                }

                "tx" -> {
                    val pathParts = uri.pathSegments

                    if (pathParts.size > 1) {
                        val chain = pathParts[0]
                        val txId = pathParts.drop(1).joinToString("/")
                        return Deeplink.Transaction(
                            accountAddress = null,
                            chain = chain,
                            txId = txId,
                            txHash = null,
                            isPushNotification = false,
                        )
                    } else {
                        return null
                    }
                }

                "view" -> {
                    val addressByChain = mutableMapOf<String, String>()
                    MBlockchain.supportedChainValues.forEach { chain ->
                        val address = uri.getQueryParameter(chain)
                        if (!address.isNullOrBlank()) {
                            val blockchain = MBlockchain.valueOf(chain)
                            if (blockchain.isValidAddress(address) || blockchain.isValidDNS(address)) {
                                addressByChain[chain] = address
                            }
                        }
                    }
                    val network =
                        if (uri.getQueryParameter("testnet") == "true") MBlockchainNetwork.TESTNET else MBlockchainNetwork.MAINNET
                    return Deeplink.View(
                        accountAddress = null,
                        network = network,
                        addressByChain = addressByChain
                    )
                }

                "nft" -> {
                    val nftAddress = uri.pathSegments.firstOrNull()?.takeIf { it.isNotBlank() } ?: return null
                    val network =
                        if (uri.getQueryParameter("testnet") == "true") MBlockchainNetwork.TESTNET else MBlockchainNetwork.MAINNET
                    return Deeplink.Nft(accountAddress = null, network = network, nftAddress = nftAddress)
                }

                else -> {
                    return null
                }
            }
        }

        private fun extractId(pathname: String, pattern: String): String? {
            val regex = Regex(pattern)
            val match = regex.find(pathname)
            return match?.groups?.get(1)?.value
        }
    }
}

fun Uri.extractSubUri(): Uri? {
    // use only host for now
    val targetHost = pathSegments.firstOrNull() ?: return null
    return Uri.Builder()
        .scheme("https")
        .authority(targetHost)
        .build()
}

fun parseWalletUrl(uri: Uri): ParsedWalletUrl? {
    if ((uri.scheme != "ton" && uri.scheme != "mtw") || uri.host != "transfer") {
        return null
    }

    val updatedUrl =
        uri.buildUpon().encodedPath(uri.encodedPath).encodedQuery(uri.encodedQuery).build()

    var address: String? = null
    val path = updatedUrl.path?.trim('/') ?: ""
    if (AddressHelpers.isValidAddress(path)) {
        address = path
    }

    var amount: String? = null
    var binary: String? = null
    var comment: String? = null
    var expiry: Long? = null
    var init: String? = null
    var jetton: String? = null
    var token: String? = null

    var hasUnsupportedParams = false

    updatedUrl.queryParameterNames.forEach { paramName ->
        val value = updatedUrl.getQueryParameter(paramName)
        if (!value.isNullOrEmpty()) {
            when (paramName) {
                "amount" -> amount = value
                "bin" -> binary = value
                "exp" -> try {
                    expiry = value.toLong()
                } catch (e: NumberFormatException) {
                    // Handle invalid amount format
                }

                "init" -> init = value
                "jetton" -> jetton = value
                "text" -> comment = value
                "token" -> token = value
                else -> hasUnsupportedParams = true
            }
        }
    }

    return ParsedWalletUrl(
        address = address ?: "",
        amount = amount,
        binary = binary,
        comment = comment,
        expiry = expiry,
        init = init,
        jetton = jetton,
        token = token,
        hasUnsupportedParams = hasUnsupportedParams
    )
}

data class ParsedWalletUrl(
    val address: String,
    val amount: String?,
    val binary: String?,
    val comment: String?,
    val expiry: Long?,
    val init: String?,
    val jetton: String?,
    val token: String?,
    val hasUnsupportedParams: Boolean
)
